### PR TITLE
Add bfloat16 support to upsample

### LINKED ISSUE #80339: BFloat16 support for upsampling on CUDA
### 🚀 The feature, motivation and pitch

Upsampling operations currently do not support bfloat16 on CUDA, as discussed in https://github.com/pytorch/pytorch/issues/57707

### Alternatives

_No response_

### Additional context

_No response_

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @ngimel

### PR DESCRIPTION
**Pull request by jamaliki**
Fixes https://github.com/pytorch/pytorch/issues/80339

### CODE REVIEW

**Comment by jamaliki** [comment]
https://github.com/pytorch/pytorch/issues/57707#issuecomment-1166667488

This was mentioned as an issue to this PR, and I can see that it happens here https://github.com/pytorch/pytorch/blob/0eee81aaadf261a9352533e81101a48fb3f44cf9/aten/src/ATen/cuda/Atomic.cuh#L35 
Although, I am not sure how to fix it exactly so I would appreciate some pointers.

**Comment by turian** [comment]
@jamaliki What needs to be done to finish this PR? Is there some specific help you need? I also have this issue: https://github.com/pytorch/pytorch/issues/84989

**Comment by jamaliki** [comment]
@turian are you able to compile and test this branch?

**Comment by turian** [comment]
@jamaliki I haven't actually ever compiled torch, particulary the .cu stuff

What I think is actually a better solution is that we write a unit test in pytorch. And check that the blfoat16 upsample results are eps close to float32 upsample. That would be the strongest approach. 

Do you have any idea in which of the `test/` folders it would be best to go?

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu [hunk 1]
```diff
@@ -306,7 +306,7 @@ static void upsample_nearest2d_out_cuda_template(
         "input tensor has spatial dimension larger than the kernel capacity");
 
     cudaStream_t stream = at::cuda::getCurrentCUDAStream();
-    AT_DISPATCH_FLOATING_TYPES_AND2(ScalarType::Half, ScalarType::Byte, input.scalar_type(), "upsample_nearest2d_out_frame", [&] {
+    AT_DISPATCH_FLOATING_TYPES_AND3(ScalarType::Half, ScalarType::BFloat16, ScalarType::Byte, input.scalar_type(), "upsample_nearest2d_out_frame", [&] {
```
[the remark this answers is not in the document]
Comment (jamaliki) [reply aten/src/ATen/native/cuda/UpSampleNearest2d.cu:309]: Hi, this was a good point. I've added it for `upsample_nearest2d_nhwc_out_frame` now as well.

**Review by ngimel — APPROVED** [review]

**Comment by eqy** [comment]
@jamaliki Could you check if including the header as was done in this WIP PR resolves your compilation issue? https://github.com/pytorch/pytorch/pull/84981/files

It seems a bit strange to me as it should be included as part of another header, but adding it seemed to help in my case

**Comment by jamaliki** [comment]
Thank you @eqy that fixed it!

**Comment by eqy** [comment]
Rebasing to retrigger CI:
@pytorchmergebot rebase

**Comment by ywalterh** [comment]
Curious about the status of this PR, is it considered abandoned? I'm thinking about looking into carrying it along. Waiting for this support too.

**Comment by jamaliki** [comment]
I'm not entirely sure why it was closed...

**Comment by eqy** [comment]
I believe the bot closes PRs automatically if there isn't any activity on them after a set time period (in this case it looks like ~1 month)

**Comment by denix56** [comment]
Is there a progress on this issue?

**Comment by turian** [comment]
@jamaliki Can you reopen a new PR?

**Comment by jamaliki** [comment]
@eqy @turian Yes I can recontinue, I need some help on why it fails compiling. Should I do a new PR?

**Comment by eqy** [comment]
Yes, please re-open this PR and I can try to see if I can reproduce the compilation issues.

**Comment by jamaliki** [comment]
@eqy I have reopened it here: https://github.com/pytorch/pytorch/pull/95159

**Pull request closed without merging**